### PR TITLE
Sycl par matmat

### PR DESCRIPTION
**Pull request by waynemitchell**
This PR includes further basic infrastructure changes to unify the GPU implementation across cuda/hip/sycl and implements the parallel matrix matrix product in sycl.

`HYPRE_CUDA_LAUNCH` and `HYPRE_SYCL_LAUNCH` macros have been unified under `HYPRE_GPU_LAUNCH` for kernel launches. The plan moving forward is that all existing kernels should be usable with either cuda/hip or sycl implementations to avoid duplicating kernel code. This will require some additional if/def-ing and small modifications in order for existing kernels to be supported in sycl.

I did some reordering of functions in several files (e.g. par_csr_matop_device.c, device_utils.c, etc.) in order to group together functions available in cuda/hip/sycl vs. functions only available in cuda/hip or only in sycl. The hope is to cut down on if/def's and make it more clear which functions are available in which implementations. The downside is lots of ugly changes that are hard to read in the git diff. I'm open to suggestions on how to address this or I can try to manually undo some of the reordering to clean up the diffs if that is preferred.

PROPOSED USER API CHANGE:
I'd propose the following change: replace `HYPRE_SetSpGemmUseCusparse` with `HYPRE_SetSpGemmUseVendor` to reflect that this actually refers to either cusparse, rocsparse, or oneMKL depending on cuda/hip/sycl backend. I know we try to avoid user API changes as much as possible, however, so I'm fine reverting this if others think that's best.

Regression testing:
- [x] Sycl build on JLSE
- [x] machine-lassen
- [x] Check hip build on crusher
- [x] machine-tux

### CODE REVIEW

File: src/parcsr_ls/par_amg_solveT.c [hunk 1]
```diff
@@ -703,10 +703,11 @@ HYPRE_Int  hypre_BoomerAMGRelaxT( hypre_ParCSRMatrix *A,
           *  Generate CSR matrix from ParCSRMatrix A
           *-----------------------------------------------------------------*/
 
+         A_CSR = hypre_ParCSRMatrixToCSRMatrixAll(A);
+         f_vector = hypre_ParVectorToVectorAll(f);
+         /* WM: If statement was in the wrong place? Need all procs to participate in comm above */
```
Comment (liruipeng) [inline src/parcsr_ls/par_amg_solveT.c]: It seems that the change is correct. But I am not 100% sure.

File: src/parcsr_mv/par_csr_matop_device.c [hunk 2]
```diff
@@ -5,435 +5,644 @@
...
-   *request_ptr = (void *) vrequest;
+   const dim3 bDim = hypre_GetDefaultDeviceBlockDimension();
+   const dim3 gDim = hypre_GetDefaultDeviceGridDimension(hypre_CSRMatrixNumRows(A_diag), "warp", bDim);
 
-   /* free */
-   hypre_TFree(send_i,     HYPRE_MEMORY_HOST);
-   hypre_TFree(recv_i,     HYPRE_MEMORY_HOST);
-   hypre_TFree(d_send_i,   HYPRE_MEMORY_DEVICE);
-   hypre_TFree(d_send_map, HYPRE_MEMORY_DEVICE);
+   /* WM: Q - need to pass variables intead of hypre_CSRMatrix...() ? Why? */
```
Comment (liruipeng) [inline src/parcsr_mv/par_csr_matop_device.c]: I am also curious

File: src/parcsr_mv/par_csr_matop_device.c [hunk 3]
```diff
@@ -1212,24 +1340,24 @@ hypre_ParCSRMatrixDropSmallEntriesDevice( hypre_ParCSRMatrix *A,
...
+                        hypre_CSRMatrixJ(A_diag), hypre_CSRMatrixData(A_diag), hypre_CSRMatrixI(A_offd),
+                        hypre_CSRMatrixData(A_offd), elmt_tols_diag, elmt_tols_offd);
    }
    if (type == 2)
    {
-      HYPRE_CUDA_LAUNCH( hypre_ParCSRMatrixDropSmallEntriesDevice_getElmtTols<2>, gDim, bDim,
-                         hypre_CSRMatrixNumRows(A_diag), tol, hypre_CSRMatrixI(A_diag),
-                         hypre_CSRMatrixJ(A_diag), hypre_CSRMatrixData(A_diag), hypre_CSRMatrixI(A_offd),
-                         hypre_CSRMatrixData(A_offd), elmt_tols_diag, elmt_tols_offd);
+      HYPRE_GPU_LAUNCH( hypre_ParCSRMatrixDropSmallEntriesDevice_getElmtTols<2>, gDim, bDim,
```
Comment (liruipeng) [inline src/parcsr_mv/par_csr_matop_device.c:1305]: It's hard for me to follow the changes in this file. Other than reordering things, name changing from CUDA to GPU, and adding new sycl code, any changes to the existing CUDA code?

File: src/parcsr_mv/par_csr_triplemat_device.c [hunk 4]
```diff
@@ -502,14 +644,15 @@ hypre_ParCSRTMatMatKTDevice( hypre_ParCSRMatrix  *A,
...
 
-
 /* C = R^{T} * A * P */
 hypre_ParCSRMatrix*
 hypre_ParCSRMatrixRAPKTDevice( hypre_ParCSRMatrix *R,
                                hypre_ParCSRMatrix *A,
                                hypre_ParCSRMatrix *P,
                                HYPRE_Int           keep_transpose )
 {
+   /* hypre_ParCSRMatrixPrint(P, "P"); */
```
Comment (liruipeng) [inline src/parcsr_mv/par_csr_triplemat_device.c]: remove?

File: src/parcsr_mv/par_csr_triplemat_device.c [hunk 5]
```diff
@@ -217,8 +268,8 @@ hypre_ParCSRTMatMatKTDevice( hypre_ParCSRMatrix  *A,
 
    HYPRE_Int num_procs;
    hypre_MPI_Comm_size(hypre_ParCSRMatrixComm(A), &num_procs);
-   //HYPRE_Int my_id;
-   //hypre_MPI_Comm_rank(hypre_ParCSRMatrixComm(A), &my_id);
+   HYPRE_Int my_id;
```
Comment (liruipeng) [inline src/parcsr_mv/par_csr_triplemat_device.c]: Is `my_id` used somewhere?

File: src/parcsr_ls/par_amg_solveT.c (hunk 1 above)
Comment (waynemitchell) [reply src/parcsr_ls/par_amg_solveT.c]: Yeah, this fixed a bug for me and also makes sense to me.

File: src/seq_mv/csr_matop_device.c [hunk 6]
```diff
@@ -11,6 +11,7 @@
  *
  *****************************************************************************/
 
+#include "_hypre_onedpl.hpp"
```
Comment (liruipeng) [inline src/seq_mv/csr_matop_device.c:14]: It's hard for me to follow the changes in this file. Other than reordering things, name changing from CUDA to GPU, and adding new sycl code, any changes to the existing CUDA code?

**Review by rfalgout — APPROVED** [review]
This looks good to me.  I will defer to the experts on the implementation details.  Thanks!

File: src/utilities/device_utils.h [hunk 7]
```diff
@@ -380,91 +507,28 @@ struct hypre_GpuMatData
...
-#if defined(HYPRE_DEBUG)
-#if defined(HYPRE_USING_CUDA)
-#define GPU_LAUNCH_SYNC { hypre_SyncComputeStream(hypre_handle()); HYPRE_CUDA_CALL( cudaGetLastError() ); }
-#elif defined(HYPRE_USING_HIP)
-#define GPU_LAUNCH_SYNC { hypre_SyncComputeStream(hypre_handle()); HYPRE_HIP_CALL( hipGetLastError() );  }
+template <typename T>
+static __device__ __forceinline__
+T read_only_load( const T *ptr )
+{
+   /* WM: Q - where/when is __CUDA_ARCH__ defined? */
```
Comment (liruipeng) [inline src/utilities/device_utils.h]: it is defined by CUDA

File: src/utilities/device_utils.h [hunk 8]
```diff
@@ -115,32 +230,13 @@
...
-      }).wait_and_throw();                                                                   \
-   }                                                                                         \
-}
-
-#endif // defined(HYPRE_USING_CUDA)
+/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+ *      macros for wrapping vendor library calls for error reporting
+ * - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
 
+/* WM: Q - any need to macro guard these definitions? */
```
Comment (liruipeng) [inline src/utilities/device_utils.h]: like putting in HYPRE_USING_CUBLAS?

File: src/utilities/device_utils.h [hunk 9]
```diff
@@ -8,6 +8,7 @@
 #ifndef HYPRE_DEVICE_UTILS_H
 #define HYPRE_DEVICE_UTILS_H
 
+/* WM: Q - do we need this macro gaurd? Where should it be placed? */
```
Comment (liruipeng) [inline src/utilities/device_utils.h]: I guess yes? Protect them for GPU build only. Is there issue leaving it here?

**Review by liruipeng — APPROVED** [review]
A lot of good changes. Looks good to me. Thanks @waynemitchell

**Comment by liruipeng** [comment]
It's indeed hard to follow the changes in `par_csr_matop_device.c, sr_matop_device.c, device_utils.c` with git. If there are no other changes than reordering, renaming and adding new sycl code, to the existing CUDA code, I am fine with these files, relying on @waynemitchell for the correctness of the new sycl/oneAPI code.

The API change will affect the users, but the new naming is way better. The change should be made. For back compatibility, can we just define `HYPRE_SetSpGemmUseCusparse` to `HYPRE_SetSpGemmUseVendor`, so we will not be yelled at?

Thanks!

File: src/utilities/device_utils.h (hunk 8 above)
Comment (waynemitchell) [reply src/utilities/device_utils.h]: Sure. I mainly ask for the sake of consistency. Other definitions (e.g. HYPRE_THRUST_CALL below) listed here are behind macro guards, so I just wanted to see whether we have a policy for this kind of thing.

File: src/utilities/HYPRE_utilities.h [hunk 10]
```diff
@@ -229,7 +229,9 @@ HYPRE_Int HYPRE_SetGPUMemoryPoolSize(HYPRE_Int bin_growth, HYPRE_Int min_bin, HY
  * HYPRE handle
  *--------------------------------------------------------------------------*/
 
-HYPRE_Int HYPRE_SetSpGemmUseCusparse( HYPRE_Int use_cusparse );
+/* Backwards compatibility with HYPRE_SetSpGemmUseCusparse() */
+#define HYPRE_SetSpGemmUseCusparse(use_vendor) HYPRE_SetSpGemmUseVendor(use_vendor)
```
Comment (waynemitchell) [inline src/utilities/HYPRE_utilities.h:233]: @liruipeng , is this what you had in mind for providing backward compatibility?

File: src/utilities/device_utils.c [hunk 11]
```diff
@@ -183,170 +223,455 @@ HYPRE_Int
...
+/* Note: d_ib is an input vector that contains row ptrs,
+ *       i.e., start positions where to put the rows in d_jb and d_ab.
+ *       The col indices in B are global indices, i.e., BigJ
+ *       of length (nrows + 1) or nrow (without the last entry, nnz) */
+/* Special cases:
+ *    if d_row_indices == NULL, it means d_row_indices=[0,1,...,nrows-1]
+ *    If col_map_offd_A == NULL, use (-1 - d_offd_j) as column id
+ *    If nrows == 1 and d_ib == NULL, it means d_ib[0] = 0 */
+HYPRE_Int
+hypreDevice_CopyParCSRRows(HYPRE_Int      nrows,
```
Comment (liruipeng) [inline src/utilities/device_utils.c]: Have this function twice in this file.

File: src/utilities/HYPRE_utilities.h (hunk 10 above)
Comment (liruipeng) [reply src/utilities/HYPRE_utilities.h:233]: yep. Thanks!

File: src/test/TEST_ij/solvers.jobs [hunk 12]
```diff
@@ -30,91 +30,91 @@
...
-mpirun -np 2 ./ij -solver 20 -cf 0.5 -rhsrand -solver_type 2 > solvers.out.10
-mpirun -np 2 ./ij -solver 20 -cf 0.5 -rhsrand -solver_type 3 > solvers.out.11
-mpirun -np 2 ./ij -solver 16 -rhsrand > solvers.out.12
-mpirun -np 2 ./ij -solver 17 -rhsrand > solvers.out.13
-mpirun -np 2 ./ij -solver 17 -rhsrand -cgs 2 > solvers.out.14
-mpirun -np 2 ./ij -solver 17 -rhsrand -cgs 2 -unroll 8 > solvers.out.15
-mpirun -np 2 ./ij -solver 17 -rhsrand -unroll 4 > solvers.out.16
-mpirun -np 2 ./ij -solver 3 -rhsrand -check_residual > solvers.out.17
-mpirun -np 2 ./ij -solver 4 -rhsrand -check_residual > solvers.out.18
+mpirun -np 2 gpu_tile_compact.sh ./ij -solver 1 -rhsrand > solvers.out.0
```
Comment (liruipeng) [inline src/test/TEST_ij/solvers.jobs]: Are you going to upload `gpu_tile_compact.sh`?

Comment (waynemitchell) [reply src/test/TEST_ij/solvers.jobs]: Ah, I didn't mean to push these changes. I've been doing regression testing on Arcticus this way. I don't think this is a good long term solution... I'm hoping that the natural behavior of MPI will eventually yield what is currently accomplished by gpu_tile_compact.sh on Arcticus.

File: src/parcsr_mv/par_csr_matop_device.c (hunk 2 above)
Comment (waynemitchell) [reply src/parcsr_mv/par_csr_matop_device.c]: After further discussion with Brian on this, it seems that this is a truth about sycl. If you have a heap-based pointer to a struct, say a hypre_CSRMatrix *A, that was allocated with a regular malloc, then it's not allowed to pass things like A->data to sycl kernel launches. Note that passing A->size or other things that aren't pointers to allocated data is OK. This is just something I'll have to keep in mind as I continue porting code.

File: src/parcsr_mv/par_csr_matop_device.c (hunk 3 above)
Comment (waynemitchell) [reply src/parcsr_mv/par_csr_matop_device.c:1305]: There should be no changes to existing code, just additions of sycl code and moving things around. I'll make sure lassen regression tests pass to double check I haven't accidentally messed with any cuda code.

File: src/seq_mv/csr_matop_device.c (hunk 6 above)
Comment (waynemitchell) [reply src/seq_mv/csr_matop_device.c:14]: Again, should be no changes to existing code.

**Comment by waynemitchell** [comment]
OK, all regression testing is done, and I think this is ready to merge.
